MoneyDataType/CurrencyConverter.cs

```diff
@@ -0,0 +1,26 @@
+using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using OrchardCore.Commerce.Abstractions;
+using OrchardCore.Commerce.Money;
+
+namespace OrchardCore.Commerce.Serialization
+{
+    internal class CurrencyConverter : JsonConverter<ICurrency>
+    {
+        public override ICurrency Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            => Currency.FromISOCode(reader.GetString());
```

Comment:
Ah! So this is problematic and why I did the 2-step resolution of currency before (which you removed). This only works for currencies in the `KnownCurrencyTable`. If a provider attempts to add new ones, they won't be able to serialize and deserialize it correctly. The problem is that deserialization can't work with dependency injection, so we need to first deserialize with the info we have, and then add the complete info when we have access to it. I'm not super-happy with the solution I had, but it worked better than this, unless I'm missing something.

Reply:
That's a good point, I'll have a think about that one today, need to get some other work down first though.

I'm happy to make to a NuGet library when it's done, but for now I think the reference as a project is better, as it'll make changes easier as the full commerce grows and becomes a module in it's own right.

Reply:
I think I might have a workable solution:

**concept** 
since Currency Providers are only relevant from within a `IMoneyService` then the money service should be responsible for providing it's providers for resolution.  Fair statement?

**solution**
`GetCurrency` in the `IMoneyService` need to change to pass the registered providers to the FromXXXX methods.
```
        public ICurrency GetCurrency(string isoCode)
            => Currency.FromISOCode(isoCode, _currencyProviders); 
```

This allows the Currency method to first check the default provider (KnownCurrencyTable) and then the extra providers. This allows us to short-circuit the find, if it's a default currency else return the first from all the passed providers.
```csharp
        public static ICurrency FromISOCode(string code, IEnumerable<ICurrencyProvider> providers = null)
        {
            var found = _defaultProvider.GetCurrency(code);
            return found ?? (providers ?? new List<ICurrencyProvider>())
                .SelectMany(p => p.Currencies)
                .FirstOrDefault(c => c.IsoCode == code);
        }
```

If currency is not used through the `IMoneyService` then there wont be any providers and the providers parameter can remain null?

Thoughts?

Reply:
I don't think that works for the deserialization issue. I don't think you can get around the fact that the deserialization cannot use anything external. Hence, you need to deserialize to a placeholder, that you can then later coerce to one of the provided currencies. I think we need to keep the idea of an unresolved currency for the custom case. You may be able however to resolve it early for the "system" ones as that's static data that's always available. Making sense?

Reply:
Oooor... We could depart from Orchard conventions for currencies and have a static registry of currencies. It comes with some caveats, but it would enable deserialization to work even for custom currencies...

Reply:
I like the idea of a registry, just not static.  We could easily create a registry and make it a singleton, thereby persisting the "default" concept.  but still allowing people to create there own and potentially overriding the provided one.

Reply:
We could then use the ContractResolver (at least in Json) to resolve the currency during deserialization.  ref: https://www.newtonsoft.com/json/help/html/DeserializeWithDependencyInjection.htm

Reply:
That might work, provided we can also do that with System.Text.Json.

Reply:
going home now, i'll give it a go.  I hope it's available in System.Text.Json since it's still only a subset.

Reply:
And if it's not, it would actually be fine to drop System.Text.Json support, as that was pretty much an experiment. Newtonsoft is fine for what we're doing here.

Reply:
Didn't need to remove System.Text.Json, and also removed 2 step resolution.  See next change set for details.